### PR TITLE
Feature: Static Expression Analysis

### PR DESCRIPTION
**Pull request by luvies**
Adds static analysis features, allowing users to scan an expression and pull out the following:

- Function calls
  - Each one includes the name, the nested path (if there is one) it took to get to it, and the arguments (which are literals, nested function calls, or a placeholder)
- Errors
  - All the errors that can be found statically are generated and returned

A set of more specific functions will come later to allow analysis of rules expressions, but that will be a separate PR.

### CODE REVIEW

File: src/info/collector.ts [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+
+  private tryResolveLiteral(
+    expression: Expression,
+    errors?: ExpressionError[],
+  ): ExpressionReturnType | FunctionCall | RuntimeValue {
+    switch (expression.type) {
+      case 'Literal':
+        return expression.value;
+      case 'Identifier':
+        return this.tryResoveFromIdentifier(expression, errors);
```
Comment (timstableford) [inline src/info/collector.ts]: typo *Resolve

File: test/info.spec.ts [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    expect(res.functionCalls).toHaveLength(0);
+    expect(res.errors).toHaveLength(1);
+
+    res = info.collect('a * (b && c(1, 2))');
+
+    expect(res.functionCalls).toHaveLength(1);
+    expect(res.errors).toHaveLength(1);
+
+    expect(res.functionCalls[0].name).toBe('c');
+    expect(res.functionCalls[0].args).toStrictEqual([1, 2]);
```
Comment (timstableford) [inline test/info.spec.ts]: Should probably check 'a' and 'b' are valid too

Comment (luvies) [reply test/info.spec.ts]: They aren't functions in the expression, so nothing is collected for them. They also exist in the context, so there are no errors for them either.

File: src/info/collector.ts (hunk 1 above)
Comment (luvies) [reply src/info/collector.ts]: goddammit

fixed

**Pull request merged**